File: cpp/check.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <cmath>

using namespace std;
using namespace std::chrono;

// Enum for shape types
enum ShapeType { CIRCLE, RECTANGLE, TRIANGLE };

// Struct for shapes used in the switch/case approach
struct ShapeStruct {
    ShapeType type;
    double dimension1;
    double dimension2;
};

// Function to calculate area using switch/case
double calculateAreaSwitchCase(const ShapeStruct& shape) {
    switch (shape.type) {
        case CIRCLE:
            return M_PI * shape.dimension1 * shape.dimension1;
        case RECTANGLE:
            return shape.dimension1 * shape.dimension2;
        case TRIANGLE:
            return 0.5 * shape.dimension1 * shape.dimension2;
        default:
            return 0.0;
    }
}

// Test function for switch/case
void testSwitchCase() {
    ShapeStruct shapes[3] = {
        {CIRCLE, 10.0, 0.0},
        {RECTANGLE, 5.0, 10.0},
        {TRIANGLE, 6.0, 8.0}
    };

    double totalArea = 0.0;
    for (int i = 0; i < 100000000; ++i) {  // Large number of iterations
        for (const auto& shape : shapes) {
            totalArea += calculateAreaSwitchCase(shape);
        }
    }
    cout << "Total Area (Switch/Case): " << totalArea << endl;
}

// Abstract Shape class for polymorphism
class Shape {
public:
    virtual ~Shape() = default;
    virtual double calculateArea() const = 0;
};

// Circle class
class Circle : public Shape {
    double radius;
public:
    Circle(double r) : radius(r) {}
    double calculateArea() const override {
        return M_PI * radius * radius;
    }
};

// Rectangle class
class Rectangle : public Shape {
    double length, width;
public:
    Rectangle(double l, double w) : length(l), width(w) {}
    double calculateArea() const override {
        return length * width;
    }
};

// Triangle class
class Triangle : public Shape {
    double base, height;
public:
    Triangle(double b, double h) : base(b), height(h) {}
    double calculateArea() const override {
        return 0.5 * base * height;
    }
};

// Test function for polymorphism
void testPolymorphism() {
    vector<Shape*> shapes = {
        new Circle(10.0),
        new Rectangle(5.0, 10.0),
        new Triangle(6.0, 8.0)
    };

    double totalArea = 0.0;
    for (int i = 0; i < 100000000; ++i) {  // Large number of iterations
        for (const auto& shape : shapes) {
            totalArea += shape->calculateArea();
        }
    }

    for (auto& shape : shapes) {
        delete shape;
    }

    cout << "Total Area (Polymorphism): " << totalArea << endl;
}

// Function to measure performance
template<typename Func, typename... Args>
void measure_perf(const string& test_name, Func func, Args&&... args) {
    auto start = high_resolution_clock::now();
    func(std::forward<Args>(args)...);
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(stop - start);
    cout << test_name << " Elapsed Time: " << duration.count() << " milliseconds" << endl;
}

int main() {
    cout << "Testing Switch/Case:" << endl;
    measure_perf("Switch/Case", testSwitchCase);

    cout << "Testing Polymorphism:" << endl;
    measure_perf("Polymorphism", testPolymorphism);

    return 0;
}

File: go/check.go
package main

// import (
//     "fmt"
//     "math"
//     "time"
// )

// // ShapeType is an enum for shape types
// type ShapeType int

// const (
//     CIRCLE ShapeType = iota
//     RECTANGLE
//     TRIANGLE
// )

// // ShapeStruct is a struct for shapes used in the switch/case approach
// type ShapeStruct struct {
//     Type       ShapeType
//     Dimension1 float64
//     Dimension2 float64
// }

// // calculateAreaSwitchCase calculates area using switch/case
// func calculateAreaSwitchCase(shape ShapeStruct) float64 {
//     switch shape.Type {
//     case CIRCLE:
//         return math.Pi * shape.Dimension1 * shape.Dimension1
//     case RECTANGLE:
//         return shape.Dimension1 * shape.Dimension2
//     case TRIANGLE:
//         return 0.5 * shape.Dimension1 * shape.Dimension2
//     default:
//         return 0.0
//     }
// }

// // testSwitchCase is a function to test switch/case approach
// func testSwitchCase() {
//     shapes := []ShapeStruct{
//         {CIRCLE, 10.0, 0.0},
//         {RECTANGLE, 5.0, 10.0},
//         {TRIANGLE, 6.0, 8.0},
//     }

//     totalArea := 0.0
//     for i := 0; i < 10000000; i++ {
//         for _, shape := range shapes {
//             totalArea += calculateAreaSwitchCase(shape)
//         }
//     }
//     fmt.Println("Total Area (Switch/Case):", totalArea)
// }

// // Shape interface for polymorphism
// type Shape interface {
//     CalculateArea() float64
// }

// // Circle struct
// type Circle struct {
//     Radius float64
// }

// // CalculateArea calculates the area of a circle
// func (c Circle) CalculateArea() float64 {
//     return math.Pi * c.Radius * c.Radius
// }

// // Rectangle struct
// type Rectangle struct {
//     Length float64
//     Width  float64
// }

// // CalculateArea calculates the area of a rectangle
// func (r Rectangle) CalculateArea() float64 {
//     return r.Length * r.Width
// }

// // Triangle struct
// type Triangle struct {
//     Base   float64
//     Height float64
// }

// // CalculateArea calculates the area of a triangle
// func (t Triangle) CalculateArea() float64 {
//     return 0.5 * t.Base * t.Height
// }

// // testPolymorphism is a function to test polymorphism approach
// func testPolymorphism() {
//     shapes := []Shape{
//         Circle{10.0},
//         Rectangle{5.0, 10.0},
//         Triangle{6.0, 8.0},
//     }

//     totalArea := 0.0
//     for i := 0; i < 10000000; i++ {
//         for _, shape := range shapes {
//             totalArea += shape.CalculateArea()
//         }
//     }
//     fmt.Println("Total Area (Polymorphism):", totalArea)
// }

// // measurePerf measures the performance of a function
// func measurePerf(testName string, f func()) {
//     start := time.Now()
//     f()
//     elapsed := time.Since(start)
//     fmt.Printf("%s Elapsed Time: %s\n", testName, elapsed)
// }

// func main() {
//     fmt.Println("Testing Switch/Case:")
//     measurePerf("Switch/Case", testSwitchCase)

//     fmt.Println("Testing Polymorphism:")
//     measurePerf("Polymorphism", testPolymorphism)
// }